core/src/main/scala/io/finch/Endpoint.scala

```diff
@@ -78,15 +78,19 @@ trait Endpoint[A] { self =>
   /**
    * Maps this endpoint to the given function `A => Future[B]`.
    */
-  final def mapAsync[B](fn: A => Future[B]): Endpoint[B] = new Endpoint[B] {
-    final def apply(input: Input): Endpoint.Result[B] = self(input) match {
-      case EndpointResult.Matched(rem, out) =>
-        EndpointResult.Matched(rem, out.flatMapF(oa => oa.traverse(a => fn(a))))
-      case _ => EndpointResult.Skipped
-    }
+  final def mapAsync[B](fn: A => Future[B]): Endpoint[B] =
+    new Endpoint[B] with (Output[A] => Future[Output[B]]) {
```

Comment:
Yeah, it's pretty mind-blowing that you can do that. I think Flavio calls this technic "merge interfaces". We did a whole bunch of optimizations for [Twitter Futures](https://github.com/twitter/util/commits/develop/util-core/src/main/scala/com/twitter/util/Future.scala) based on that (and other ideas). Results were outstanding.